compiler/src/org.graalvm.compiler.lir.aarch64/src/org/graalvm/compiler/lir/aarch64/AArch64AtomicMove.java

```diff
@@ -56,61 +60,82 @@
 
             Register address = asRegister(addressValue);
             Register result = asRegister(resultValue);
             Register newVal = asRegister(newValue);
-            if (useBarriers) {
-                masm.dmb(AArch64Assembler.BarrierKind.ANY_ANY);
-                emitCompareAndSwap(masm, size, address, result, newVal, false, false);
-                masm.dmb(AArch64Assembler.BarrierKind.ANY_ANY);
-            } else {
-                emitCompareAndSwap(masm, size, address, result, newVal, true, true);
-            }
-        }
+            Register expected = asRegister(expectedValue);
+
+            /**
+             * Determining whether acquire and/or release semantics are needed.
+             */
+            boolean acquire = (memoryOrder.postReadBarriers & (LOAD_LOAD | LOAD_STORE)) != 0;
+            boolean release = (memoryOrder.preWriteBarriers & (LOAD_STORE | STORE_STORE)) != 0;
 
-        private void emitCompareAndSwap(AArch64MacroAssembler masm, int size, Register address, Register result, Register newVal, boolean acquire, boolean release) {
             if (AArch64LIRFlagsVersioned.useLSE(masm.target.arch)) {
-                Register expected = asRegister(expectedValue);
-                masm.mov(size, result, expected);
-                masm.cas(size, result, newVal, address, acquire, release);
+                masm.mov(Math.max(memAccessSize, 32), result, expected);
+                masm.cas(memAccessSize, result, newVal, address, acquire, release);
                 AArch64Compare.gpCompare(masm, resultValue, expectedValue);
             } else {
                 // We could avoid using a scratch register here, by reusing resultValue for the
                 // stlxr success flag and issue a mov resultValue, expectedValue in case of success
                 // before returning.
+
+                /*
+                 * If both an acquire and release are requested, a full barrier is expected, and no
+                 * instructions should reordered across the atomic operations. Unfortunately, since
+                 * ldaxr and stlxr implement "half" barriers, then it is possible for instructions
+                 * on either side of the atomic operation to be reordered. For example,
+                 *
+                 * A -> ldaxr -> stlxr -> B
+                 *
+                 * can execute as:
+                 *
+                 * ldaxr -> B -> A -> stlxr
+                 *
+                 * Note that only dmb is needed to prevent instructions before the atomic operation
+                 * from executing too late. All instructions after the atomic operation are ordered
+                 * after the ldaxr and, if it occurs, will also see updated value since it is a
+                 * stxr.
+                 */
+                boolean fullBarrier = acquire && release;
+                if (fullBarrier) {
+                    masm.dmb(AArch64Assembler.BarrierKind.ANY_ANY);
+                }
+
```

Comment:
Hi @teshull thanks for looking at the atomic patches.  For the fullbarrier mode, do you mean Graal will always insert the "dmb" before CAS, even it is generated for the Java Unsafe intrinisics?  I added the flag `useBarrier` to distinguish the JAVA atomics (should be` ldaxr+stlxr`) and vm internal atomics which needs conservative barriers (see https://github.com/oracle/graal/pull/2181).  
Regarding to the vm internal CAS, here is a related discussing thread: https://patchwork.kernel.org/project/linux-arm-kernel/patch/1391516953-14541-1-git-send-email-will.deacon@arm.com/

Regarding to this patch, it looks the final memory barriers are not the same with the above suggested ones.  WDYT? Please correct me if I have any misunderstanding. Thanks!

Reply:
Hi @XiaohongGong. The code here is out of date. Here is what I am planning to merge:
```
/*
* Because the store is only conditionally emitted, a dmb is needed for performing a
* release.
*
* Furthermore, even if the stlxr is emitted, if both acquire and release semantics
* are required, then a dmb is anyways needed to ensure that the instruction
* sequence:
*
* A -> ldaxr -> stlxr -> B
*
* can not be executed as:
*
* ldaxr -> B -> A -> stlxr
*/
if (release) {
masm.dmb(AArch64Assembler.BarrierKind.ANY_ANY);
}

Register scratch = asRegister(scratchValue);
Label retry = new Label();
Label fail = new Label();
masm.bind(retry);
masm.loadExclusive(memAccessSize, result, address, acquire);
AArch64Compare.gpCompare(masm, resultValue, expectedValue);
masm.branchConditionally(AArch64Assembler.ConditionFlag.NE, fail);
masm.storeExclusive(memAccessSize, scratch, newVal, address, false);
// if scratch == 0 then write successful, else retry.
masm.cbnz(32, scratch, retry);
masm.bind(fail);
```

I realized that since the store is only conditionally emitted, it was not possible to tie any release semantics to it. Anyways, I think for a cas, since its is both an acquire & release, that a dmb is necessary due to the link you provided.

Reply:
So to prevent the following reordering:
```
* A -> ldaxr -> stlxr -> B
*
* can not be executed as:
*
* ldaxr -> B -> A -> stlxr
```
Do you think a DMB after `stlxr` is needed? Also due to the discussions inside the thread I pointed, it seems that the final memory barrier mode should be: `ldxr, stlxr, dmb` , which is different from what you mentioned above `(dmb, ldaxr, stxr)`. Any others that I was missing or misunderstanding? Thanks!

Reply:
I believe `dmb, ldaxr, stxr` should act a full barrier the same way as 'ldxr, stlxr, dmb`. However, we can't use `ldxr, stlxr, dmb`, as the store is only conditionally executed.

Within `dmb, ldaxr, stxr`, the dmb prevent all prior instructions from executing. Likewise, the ldaxr prevents all subsequent instructions from execution

Reply:
Thanks for more explanation! It seems reasonable.